File: packages/fitahead_character/lib/src/manifest.dart
import 'dart:convert';

import 'vec3.dart';

/// Parsed form of `assets/models/character_manifest.json`.
///
/// The manifest is generated next to the GLBs by `tools/gen_character.py`, so
/// joint indices, morph target names and camera framings can never drift out of
/// sync with the meshes the app actually loads.
class CharacterManifest {
  const CharacterManifest({
    required this.version,
    required this.faceTexture,
    required this.presets,
  });

  factory CharacterManifest.fromJson(Map<String, dynamic> json) {
    final presets = <String, CharacterPreset>{};
    (json['presets'] as Map<String, dynamic>).forEach((key, value) {
      presets[key] = CharacterPreset.fromJson(value as Map<String, dynamic>);
    });
    return CharacterManifest(
      version: json['version'] as int,
      faceTexture: json['faceTexture'] as String,
      presets: Map.unmodifiable(presets),
    );
  }

  factory CharacterManifest.parse(String source) =>
      CharacterManifest.fromJson(jsonDecode(source) as Map<String, dynamic>);

  final int version;

  /// Filename of the placeholder face texture, relative to the manifest.
  final String faceTexture;

  final Map<String, CharacterPreset> presets;

  CharacterPreset preset(String id) {
    final p = presets[id];
    if (p == null) {
      throw ArgumentError.value(
        id,
        'id',
        'unknown preset; available: ${presets.keys.join(', ')}',
      );
    }
    return p;
  }
}

/// One buildable character: the male or the female body.
class CharacterPreset {
  const CharacterPreset({
    required this.id,
    required this.sex,
    required this.file,
    required this.height,
    required this.meshes,
    required this.face,
    required this.joints,
    required this.morphTargets,
    required this.focusViews,
    required this.focusFovDeg,
    required this.archetypes,
    required this.humanoidBones,
    required this.absFatOcclusion,
    required this.fatSplit,
    required this.stats,
  });

  factory CharacterPreset.fromJson(Map<String, dynamic> json) {
    return CharacterPreset(
      id: json['id'] as String,
      sex: json['sex'] as String,
      file: json['file'] as String,
      height: (json['height'] as num).toDouble(),
      meshes: List<String>.unmodifiable(
        (json['meshes'] as List).cast<String>(),
      ),
      face: FaceSlot.fromJson(json['face'] as Map<String, dynamic>),
      joints: List<JointInfo>.unmodifiable(
        (json['joints'] as List)
            .map((e) => JointInfo.fromJson(e as Map<String, dynamic>)),
      ),
      morphTargets: List<MorphTargetInfo>.unmodifiable(
        (json['morphTargets'] as List)
            .map((e) => MorphTargetInfo.fromJson(e as Map<String, dynamic>)),
      ),
      focusViews: List<FocusView>.unmodifiable(
        (json['focusViews'] as List)
            .map((e) => FocusView.fromJson(e as Map<String, dynamic>)),
      ),
      focusFovDeg: (json['focusFovDeg'] as num).toDouble(),
      archetypes: List<BodyArchetype>.unmodifiable(
        (json['archetypes'] as List)
            .map((e) => BodyArchetype.fromJson(e as Map<String, dynamic>)),
      ),
      humanoidBones: Map<String, String>.unmodifiable(
        (json['humanoidBones'] as Map<String, dynamic>).cast<String, String>(),
      ),
      absFatOcclusion: (json['absFatOcclusion'] as num).toDouble(),
      fatSplit: FatSplit.fromJson(json['fatSplit'] as Map<String, dynamic>),
      stats: PresetStats.fromJson(json['stats'] as Map<String, dynamic>),
    );
  }

  final String id;
  final String sex;

  /// GLB filename, relative to the manifest.
  final String file;

  /// Total body height in metres, at rest with every morph at zero.
  final double height;

  final List<String> meshes;
  final FaceSlot face;
  final List<JointInfo> joints;
  final List<MorphTargetInfo> morphTargets;
  final List<FocusView> focusViews;
  final double focusFovDeg;

  /// Named starting body types, lean to overweight.
  final List<BodyArchetype> archetypes;

  /// `EXT_skeleton_humanoid` bone role -> our node name.
  final Map<String, String> humanoidBones;

  /// How much body fat hides abdominal definition. Visible abs are a
  /// body-composition fact as much as a training one.
  final double absFatOcclusion;

  final FatSplit fatSplit;

  final PresetStats stats;

  BodyArchetype? archetype(String id) {
    for (final a in archetypes) {
      if (a.id == id) return a;
    }
    return null;
  }

  /// Morph groups that grow with training.
  List<MorphTargetInfo> get muscleGroups => [
        for (final m in morphTargets)
          if (m.isMuscle) m
      ];

  /// Morph groups driven by body composition.
  List<MorphTargetInfo> get fatGroups => [
        for (final m in morphTargets)
          if (m.isFat) m
      ];

  /// Morph target names in GLB order — the order `morphTargetInfluences`
  /// expects when a renderer does not expose a name lookup.
  List<String> get morphNames => [for (final m in morphTargets) m.name];

  MorphTargetInfo? morph(String name) {
    for (final m in morphTargets) {
      if (m.name == name) return m;
    }
    return null;
  }

  FocusView? view(String id) {
    for (final v in focusViews) {
      if (v.id == id) return v;
    }
    return null;
  }

  /// Views that map onto a trainable muscle group, in manifest order.
  List<FocusView> get trainableViews => [
        for (final v in focusViews)
          if (v.morph != null) v
      ];
}

/// Where the user's own face drawing goes.
class FaceSlot {
  const FaceSlot({
    required this.mesh,
    required this.material,
    required this.uvLayout,
    required this.defaultTexture,
  });

  factory FaceSlot.fromJson(Map<String, dynamic> json) => FaceSlot(
        mesh: json['mesh'] as String,
        material: json['material'] as String,
        uvLayout: json['uvLayout'] as String,
        defaultTexture: json['defaultTexture'] as String,
      );

  /// Name of the mesh carrying the face plate.
  final String mesh;

  /// Name of the material whose base colour texture to swap.
  final String material;

  /// How the plate's UVs are laid out; `disc-inscribed-in-square` means a
  /// square image is shown as the largest circle that fits inside it.
  final String uvLayout;

  final String defaultTexture;
}

class JointInfo {
  const JointInfo({
    required this.index,
    required this.name,
    required this.parent,
    required this.rest,
    required this.humanoid,
  });

  factory JointInfo.fromJson(Map<String, dynamic> json) => JointInfo(
        index: json['index'] as int,
        name: json['name'] as String,
        parent: json['parent'] as String?,
        rest: Vec3.fromList(json['rest'] as List),
        humanoid: json['humanoid'] as String?,
      );

  final int index;
  final String name;
  final String? parent;

  /// Rest-pose position in model space, in metres.
  final Vec3 rest;

  /// `EXT_skeleton_humanoid` bone role, or null for joints outside that set
  /// (only `Root` today). Lets humanoid animation be remapped by role rather
  /// than by node index.
  final String? humanoid;
}

/// Mesh facts about a generated preset.
class PresetStats {
  const PresetStats({
    required this.vertices,
    required this.triangles,
    required this.joints,
    required this.morphTargets,
    required this.muscleGroups,
    required this.boundsMin,
    required this.boundsMax,
    required this.measuredHeight,
  });

  factory PresetStats.fromJson(Map<String, dynamic> json) => PresetStats(
        vertices: json['vertices'] as int,
        triangles: json['triangles'] as int,
        joints: json['joints'] as int,
        morphTargets: json['morphTargets'] as int,
        muscleGroups: json['muscleGroups'] as int,
        boundsMin: _vec(json['boundsMin'] as List),
        boundsMax: _vec(json['boundsMax'] as List),
        measuredHeight: (json['measuredHeight'] as num).toDouble(),
      );

  static List<double> _vec(List<dynamic> v) =>
      List<double>.unmodifiable(v.map((e) => (e as num).toDouble()));

  final int vertices;
  final int triangles;
  final int joints;
  final int morphTargets;
  final int muscleGroups;

  /// Axis-aligned bounds at rest, in metres. `boundsMin[1]` is the sole: the
  /// character stands on y = 0, so the app can place it on a floor directly.
  final List<double> boundsMin;
  final List<double> boundsMax;

  /// Height of the actual mesh. Should equal the preset's stated height — it is
  /// asserted in the tests because it silently did not for a while.
  final double measuredHeight;
}

/// Sex-typical split of body fat between the two deposition patterns.
class FatSplit {
  const FatSplit({required this.android, required this.gynoid});

  factory FatSplit.fromJson(Map<String, dynamic> json) => FatSplit(
        android: (json['android'] as num).toDouble(),
        gynoid: (json['gynoid'] as num).toDouble(),
      );

  /// Abdominal ("apple") share — predominates in males.
  final double android;

  /// Gluteofemoral ("pear") share — predominates in females.
  final double gynoid;
}

/// A named starting body type — one point in morph space, not a separate asset.
class BodyArchetype {
  const BodyArchetype({
    required this.id,
    required this.labelKo,
    required this.labelEn,
    required this.muscleLevel,
    required this.fatLevel,
    required this.noteKo,
    required this.weights,
  });

  factory BodyArchetype.fromJson(Map<String, dynamic> json) => BodyArchetype(
        id: json['id'] as String,
        labelKo: json['labelKo'] as String,
        labelEn: json['labelEn'] as String,
        muscleLevel: (json['muscleLevel'] as num).toDouble(),
        fatLevel: (json['fatLevel'] as num).toDouble(),
        noteKo: json['noteKo'] as String,
        weights: Map<String, double>.unmodifiable({
          for (final e in (json['weights'] as Map<String, dynamic>).entries)
            e.key: (e.value as num).toDouble(),
        }),
      );

  final String id;
  final String labelKo;
  final String labelEn;

  /// 0..1 training level applied across every muscle group.
  final double muscleLevel;

  /// 0..1 body fat, split between the two deposition patterns.
  final double fatLevel;

  final String noteKo;

  /// Ready-to-apply morph weights.
  final Map<String, double> weights;
}

/// A blend shape: one muscle group that can grow.
class MorphTargetInfo {
  const MorphTargetInfo({
    required this.index,
    required this.name,
    required this.labelKo,
    required this.labelEn,
    required this.maxDisplacement,
    required this.kind,
    required this.muscleKo,
    required this.muscleEn,
  });

  factory MorphTargetInfo.fromJson(Map<String, dynamic> json) =>
      MorphTargetInfo(
        index: json['index'] as int,
        name: json['name'] as String,
        labelKo: json['labelKo'] as String,
        labelEn: json['labelEn'] as String,
        maxDisplacement: (json['maxDisplacement'] as num).toDouble(),
        kind: json['kind'] as String,
        muscleKo: json['muscleKo'] as String,
        muscleEn: json['muscleEn'] as String,
      );

  final int index;
  final String name;
  final String labelKo;
  final String labelEn;

  /// How far the surface moves, in metres, at weight 1.0.
  final double maxDisplacement;

  /// `muscle` grows with training, `fat` comes from body composition, and
  /// `composite` is derived from the other groups rather than set directly.
  final String kind;

  /// Anatomical name of the structure, for a UI that wants to teach as well as
  /// motivate ("가슴" / "대흉근").
  final String muscleKo;
  final String muscleEn;

  bool get isMuscle => kind == 'muscle';
  bool get isFat => kind == 'fat';
}

/// A camera framing for the whole body or for one body part.
///
/// [yawDeg] is measured around +Y with 0 facing the character's front (+Z), and
/// [pitchDeg] is degrees above the horizon.
class FocusView {
  const FocusView({
    required this.id,
    required this.labelKo,
    required this.labelEn,
    required this.target,
    required this.yawDeg,
    required this.pitchDeg,
    required this.distance,
    required this.framingRadius,
    required this.morph,
  });

  factory FocusView.fromJson(Map<String, dynamic> json) => FocusView(
        id: json['id'] as String,
        labelKo: json['labelKo'] as String,
        labelEn: json['labelEn'] as String,
        target: Vec3.fromList(json['target'] as List),
        yawDeg: (json['yawDeg'] as num).toDouble(),
        pitchDeg: (json['pitchDeg'] as num).toDouble(),
        distance: (json['distance'] as num).toDouble(),
        framingRadius: (json['framingRadius'] as num).toDouble(),
        morph: json['morph'] as String?,
      );

  final String id;
  final String labelKo;
  final String labelEn;

  /// Point the camera looks at, in model space.
  final Vec3 target;

  final double yawDeg;
  final double pitchDeg;

  /// Nominal camera distance for the manifest's own field of view and a square
  /// viewport. Prefer deriving it from [framingRadius] at runtime so the shot
  /// survives a different aspect ratio.
  final double distance;

  /// Radius of the sphere this view is meant to fill.
  final double framingRadius;

  /// Morph target this view showcases, or null for whole-body views.
  final String? morph;
}

File: tools/fitahead_gen/manifest.py
"""Runtime manifest: what the Flutter app needs to know about a generated GLB.

The app should never hardcode a joint index, a morph index, or a camera
position — everything it needs to focus on a body part, grow it, or pick a
starting body type is described here, regenerated alongside the mesh so the two
can never drift apart.
"""

import math

from . import anthro
from .rig import HUMANOID_BONES

#: Vertical field of view the framing distances assume, in degrees.
FOCUS_FOV = 32.0


def _distance_for(radius, fov=FOCUS_FOV):
    """Camera distance at which a sphere of `radius` is tangent to the frustum.

    sin rather than tan: the limit is the angle the sphere subtends, not the
    half-width of a plane through its centre. The Dart FocusCamera repeats this
    with the viewport's own aspect ratio; this value is the square-viewport case.
    """
    return radius / math.sin(math.radians(fov) / 2.0)


def _view(view_id, label_ko, label_en, target, radius, height, groups,
          yaw=0.0, pitch=0.0, morph=None):
    # A part view is framed on the rest pose, but the user will see it at every
    # stage of growth. Reserve room for the group's own swell plus the overall
    # lean-mass shape, so a maxed-out biceps does not overflow its own close-up.
    if morph is not None:
        headroom = groups[morph].amount + groups["bulk"].amount
        radius += headroom * height * 1.7
    return {
        "id": view_id,
        "labelKo": label_ko,
        "labelEn": label_en,
        "target": [round(v, 5) for v in target],
        "yawDeg": yaw,
        "pitchDeg": pitch,
        "distance": round(_distance_for(radius), 5),
        "framingRadius": round(radius, 5),
        "morph": morph,
    }


def build_focus_views(p, rig, groups):
    """Camera framings for the whole body and for each trainable muscle.

    Yaw is degrees around +Y with 0 facing the character's front (+Z); pitch is
    degrees above the horizon. Distance is metres from the target.

    Angles are chosen so the muscle in question faces the camera: the triceps
    sits on the back of the arm, so its view is behind the shoulder line, and the
    hamstrings view is behind the thigh. A single front-on shot per body part
    would hide half of these.
    """
    h = p.height
    sw = p.shoulder_half_w * h
    head_y = rig.world("Head")[1]

    def y(fraction):
        return fraction * h

    def midpoint(a, b):
        """Midpoint of a bone, taken from the rig.

        Derived rather than computed from landmark heights, because the arms are
        abducted: in an A-pose the middle of the humerus is neither at
        `(shoulder_y + elbow_y) / 2` nor on the body's centre line.
        """
        pa, pb = rig.world(a), rig.world(b)
        return tuple((pa[i] + pb[i]) * 0.5 for i in range(3))

    upperarm_mid = midpoint("UpperArm_L", "Forearm_L")
    forearm_mid = midpoint("Forearm_L", "Hand_L")
    thigh_mid = midpoint("Thigh_L", "Shin_L")
    shin_mid = midpoint("Shin_L", "Foot_L")

    def view(*args, **kwargs):
        return _view(*args, height=h, groups=groups, **kwargs)

    return [
        view("full_body", "전신", "Full body", (0.0, y(0.50), 0.0), 0.56 * h),
        view("upper_body", "상체", "Upper body", (0.0, y(0.70), 0.0), 0.30 * h),
        view("lower_body", "하체", "Lower body", (0.0, y(0.30), 0.0), 0.30 * h),
        view("face", "얼굴", "Face", (0.0, head_y, 0.0), 0.150 * h),

        # -- upper body -----------------------------------------------------
        view("traps", "승모근", "Traps",
             (0.0, y(p.shoulder_y + 0.005), 0.0), 0.215 * h,
             yaw=18.0, pitch=26.0, morph="traps"),
        view("delts", "어깨", "Shoulders",
             (sw * 0.72, y(p.shoulder_y), 0.0), 0.200 * h,
             yaw=34.0, pitch=12.0, morph="delts"),
        view("pecs", "가슴", "Chest",
             (0.0, y(p.chest_y + 0.015), 0.018 * h), 0.205 * h, morph="pecs"),
        view("lats", "등", "Back",
             (0.0, y((p.chest_y + p.waist_y) * 0.5), -0.018 * h), 0.240 * h,
             yaw=180.0, morph="lats"),

        # -- arms -----------------------------------------------------------
        view("biceps", "이두", "Biceps",
             upperarm_mid, 0.150 * h, yaw=30.0, morph="biceps"),
        view("triceps", "삼두", "Triceps",
             upperarm_mid, 0.150 * h, yaw=150.0, morph="triceps"),
        view("forearms", "전완", "Forearms",
             forearm_mid, 0.130 * h, yaw=30.0, morph="forearms"),

        # -- core -----------------------------------------------------------
        view("abs", "복근", "Abs",
             (0.0, y((p.crotch_y + p.chest_y) * 0.5), 0.018 * h), 0.185 * h,
             morph="abs"),
        view("obliques", "옆구리", "Obliques",
             (0.0, y(p.waist_y - 0.01), 0.0), 0.190 * h,
             yaw=68.0, morph="obliques"),

        # -- lower body -----------------------------------------------------
        view("glutes", "엉덩이", "Glutes",
             (0.0, y(p.hip_y - 0.005), -0.018 * h), 0.205 * h,
             yaw=172.0, pitch=-8.0, morph="glutes"),
        view("quads", "허벅지 앞", "Quads",
             thigh_mid, 0.195 * h, yaw=26.0, morph="quads"),
        view("hams", "허벅지 뒤", "Hamstrings",
             thigh_mid, 0.195 * h, yaw=154.0, morph="hams"),
        view("calves", "종아리", "Calves",
             shin_mid, 0.155 * h, yaw=154.0, morph="calves"),
    ]


#: How much high body fat hides abdominal definition.
#:
#: Visible abs are as much a body-composition fact as a training one — the
#: rectus abdominis can be well developed and completely invisible under a layer
#: of subcutaneous fat. Without this the character would show a six-pack on an
#: overweight body, which every user would read as wrong.
ABS_FAT_OCCLUSION = 0.90


def build_archetypes(p, groups):
    """Named starting body types, as explicit morph weights.

    These are the answer to "can it do 마른 / 평범 / 근육질": all of them are one
    mesh at different points in morph space. Nothing here is a separate asset.

    `muscle` is applied to every trainable group; `fat` is split between the two
    deposition patterns using the sex-typical ratio from `anthro.FAT_SPLIT`.
    """
    android_share, gynoid_share = anthro.FAT_SPLIT[p.sex]
    muscle_names = [g.name for g in groups.values() if g.kind == "muscle"]

    def weights(muscle, fat):
        w = {name: round(muscle, 4) for name in muscle_names}
        # abs are muscle you can only see at low body fat
        w["abs"] = round(muscle * (1.0 - fat * ABS_FAT_OCCLUSION), 4)
        w["bulk"] = round(muscle * 0.85, 4)
        w["fat_android"] = round(fat * android_share, 4)
        w["fat_gynoid"] = round(fat * gynoid_share, 4)
        return w

    spec = [
        ("lean", "마른 체형", "Lean", 0.00, 0.08,
         "훈련 이력이 없고 체지방이 낮은 몸. 기본 메시 그대로입니다."),
        ("average", "평범한 체형", "Average", 0.30, 0.42,
         "일반 성인 평균. 인체측정 평균 둘레에 해당합니다."),
        ("athletic", "운동하는 체형", "Athletic", 0.62, 0.24,
         "꾸준히 운동한 몸. 복근이 드러나기 시작합니다."),
        ("muscular", "근육질 체형", "Muscular", 0.92, 0.16,
         "훈련량이 많은 몸. 인체측정 trained 둘레에 해당합니다."),
        ("overweight", "과체중 체형", "Overweight", 0.22, 0.86,
         "체지방이 높은 몸. 근육이 있어도 복근은 보이지 않습니다."),
    ]
    return [
        {
            "id": key,
            "labelKo": ko,
            "labelEn": en,
            "muscleLevel": muscle,
            "fatLevel": fat,
            "noteKo": note,
            "weights": weights(muscle, fat),
        }
        for key, ko, en, muscle, fat, note in spec
    ]


def build_manifest(character, filename, mesh_names, face_texture):
    p, rig = character.p, character.rig
    groups = {g.name: g for g in character.morph_groups}

    return {
        "id": p.name,
        "sex": p.sex,
        "file": filename,
        "height": round(p.height, 4),
        "upAxis": "Y",
        "frontAxis": "+Z",
        "meshes": mesh_names,
        "baseState": "lean",
        "face": {
            "mesh": "Face",
            "material": "Face",
            "uvLayout": "disc-inscribed-in-square",
            "defaultTexture": face_texture,
            "note": "Replace this texture with the user's own drawing; "
                    "the plate's UVs fill the full 0..1 square.",
        },
        "joints": [
            {"index": b.index, "name": b.name, "parent": b.parent,
             "rest": [round(v, 5) for v in b.world],
             "humanoid": HUMANOID_BONES.get(b.name)}
            for b in rig.bones
        ],
        "humanoidBones": {
            humanoid: node for node, humanoid in HUMANOID_BONES.items()
        },
        "morphTargets": [
            {"index": i, "name": g.name, "kind": g.kind,
             "labelKo": g.label_ko, "labelEn": g.label_en,
             "muscleKo": g.muscle_ko, "muscleEn": g.muscle_en,
             "maxDisplacement": round(g.amount * p.height, 5)}
            for i, g in enumerate(character.morph_groups)
        ],
        "archetypes": build_archetypes(p, groups),
        "absFatOcclusion": ABS_FAT_OCCLUSION,
        "fatSplit": {"android": anthro.FAT_SPLIT[p.sex][0],
                     "gynoid": anthro.FAT_SPLIT[p.sex][1]},
        "focusViews": build_focus_views(p, rig, groups),
        "focusFovDeg": FOCUS_FOV,
        "stats": character.stats(),
    }
